src/Compiler.cpp

```diff
@@ -201,108 +287,139 @@ TfLiteNode g_nodes[)"
+    g_tensors[i].quantization = tensors[i].quantization;
+    if (tensors[i].quantization.type == kTfLiteAffineQuantization) {
+      TfLiteAffineQuantization const* quant = ((TfLiteAffineQuantization const*)(tensors[i].quantization.params));
+      g_tensors[i].params.scale = quant->scale->data[0];
+      g_tensors[i].params.zero_point = quant->zero_point->data[0];
     }
   }
-  wr << "\n";
+)";
   for (size_t i = 0; i < registrations_.size(); i++) {
-    auto opName = tflite::EnumNameBuiltinOperator(registrations_[i].code);
-    wr << "  g_registrations[" << i << "] = tflite::ops::micro::Register_"
-       << opName << "();\n";
-  }
-  wr << "\n";
-  for (size_t i = 0; i < nodes_.size(); i++) {
-    std::string nodeI = "  g_nodes[" + std::to_string(i) + "].";
-    wr << nodeI << "inputs = (TfLiteIntArray*)&" << prefix_ << "inputs" << i
-       << ";\n";
-    wr << nodeI << "outputs = (TfLiteIntArray*)&" << prefix_ << "outputs" << i
-       << ";\n";
-    wr << nodeI << "temporaries = nullptr;\n";
-    // TODO: Is this cast safe or does the data need to be non-const?
-    if (nodes_[i].node.builtin_data) {
-      wr << nodeI
-         << "builtin_data = const_cast<void*>(static_cast<const void*>(&"
-         << prefix_ << "opdata" << i << "));\n";
+    std::string opName;
+    if (registrations_[i].code == tflite::BuiltinOperator_CUSTOM) {
+      opName = registrations_[i].custom_name;
     } else {
-      wr << nodeI << "builtin_data = nullptr;\n";
+      opName = tflite::EnumNameBuiltinOperator(registrations_[i].code);
     }
-    wr << nodeI << "custom_initial_data = nullptr;\n";
-    wr << nodeI << "custom_initial_data_size = 0;\n";
-    wr << nodeI << "delegate = nullptr;\n";
+    wr << "  g_registrations[OP_" << opName << "] = tflite::ops::micro::Register_"
+        << opName << "();\n";
   }
   wr << "\n";
-  for (size_t i = 0; i < nodes_.size(); i++) {
-    auto &node = nodes_[i].node;
-    auto regI = nodes_[i].regIndex;
-    if (registrations_[regI].reg->init) {
-      std::string nodeStr = "g_nodes[" + std::to_string(i) + "]";
-      std::string ptrArg = node.builtin_data
-                               ? "(const char *)" + nodeStr + ".builtin_data"
-                               : "nullptr";
-
-      // Length arg should be zero according to doc.
-      wr << "  " << nodeStr << ".user_data = g_registrations[" << regI
-         << "]->init(&g_ctx, " << ptrArg << ", 0);\n";
+  wr << "  for(size_t i = 0; i < " << nodes_.size() << R"(; ++i) {
+    g_nodes[i].inputs = nodes[i].inputs;
+    g_nodes[i].outputs = nodes[i].outputs;
+    g_nodes[i].temporaries = nullptr;
+    g_nodes[i].builtin_data = nodes[i].builtin_data;
+    g_nodes[i].custom_initial_data = nullptr;
+    g_nodes[i].custom_initial_data_size = 0;
+    g_nodes[i].delegate = nullptr;
+    if (g_registrations[nodes[i].used_op_index]->init) {
+      g_nodes[i].user_data = g_registrations[nodes[i].used_op_index]->init(&g_ctx, (const char*)g_nodes[i].builtin_data, 0);
```

Comment:
Ah, this is why you asked about the zero for custom ops. It could be fine with the tested ones now, but since the interface demands the size here, would should pass it. This should not cause large overhead anyways if we keep a simple const array.

Reply:
see bug #17